rbx_types/src/terrain.rs

```diff
@@ -0,0 +1,548 @@
+    }
+
+    /// Finds a `Voxel` at the given position in this `Chunk`,
+    /// returning a mutable reference to it if it exists.
+    #[inline]
+    pub fn get_voxel_mut(&mut self, position: &VoxelCoordinates) -> Option<&mut Voxel> {
+        self.grid.get_mut(position)
+    }
+
+    /// Writes (or overwrites) a `Chunk` at the given position to this `Terrain`.
+    #[inline]
+    pub fn write_voxel(&mut self, position: &VoxelCoordinates, voxel: Voxel) {
+        self.grid.insert(*position, voxel);
+    }
+}
+
+impl TerrainSerializer for Chunk {
+    fn encode(&self) -> Vec<u8> {
+        // ~256 bytes if all voxels are air/base mat with maximum count. Double it
+        let mut data = Vec::<u8>::with_capacity(512);
+
+        let base_voxel: Voxel = Voxel {
+            solid_occupancy: 1.0,
+            water_occupancy: 0.0,
+            material: self.base_material,
+        };
+
+        let mut pos_cursor = VoxelCoordinates::default();
+        let mut run_length_cursor = (0u16, &base_voxel);
+        for y in 0..CHUNK_SIZE {
+            pos_cursor.0.y = y;
+            for z in 0..CHUNK_SIZE {
+                pos_cursor.0.z = z;
+                for x in 0..CHUNK_SIZE {
+                    pos_cursor.0.x = x;
+
+                    let grabbed_voxel = match self.grid.get(&pos_cursor) {
+                        Some(v) => v,
+                        _ => &base_voxel,
+                    };
+
+                    if run_length_cursor.0 == 0 {
+                        // We don't add 1 here, next if statement does it.
+                        run_length_cursor.1 = grabbed_voxel;
+                    }
+                    if grabbed_voxel == run_length_cursor.1 {
+                        if run_length_cursor.0 < 0xFF {
+                            run_length_cursor.0 += 1;
+                            continue;
+                        } else {
+                            // Properly reset the run-length if we hit the max.
+                            data.extend(grabbed_voxel.encode_run_length(run_length_cursor.0 + 1));
+                            run_length_cursor.0 = 0;
+                            continue;
+                        }
+                    }
+
+                    data.extend(run_length_cursor.1.encode_run_length(run_length_cursor.0));
+                    run_length_cursor.0 = 1;
+                    run_length_cursor.1 = grabbed_voxel;
+                }
+            }
+        }
+
+        // We might have a bit of leftovers after that loop.
+        if run_length_cursor.0 > 0 {
+            data.extend(run_length_cursor.1.encode_run_length(run_length_cursor.0));
+        }
+        data
+    }
+}
+
+/// A container allowing the modification, encoding, and decoding of the
+/// `SmoothGrid` data used by Roblox's `Terrain` object.
+#[derive(Debug, Clone, Default, PartialEq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(transparent)
+)]
+pub struct Terrain {
```

Comment:
This should be named `SmoothGrid` to keep it in theme with the other BinaryString data types.